accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+        inferred type of the expression `e1` in context `?`.
+
+    *   If `P` is `Set<Ps>` then the inferred set element type of `element` is
+        the inferred type of the expression `e1` in context `Ps`.
+
+        **TODO: what if `P` is `Iterable<R>`? Is this allowed?:**
+
+        ```dart
+        Iterable<int> = {123 as dynamic}; // Infers Set<int>?
+        ```
+
+*   If `element` is a `mapEntry` `ek: ev`:
+
+    *   If `P` is `?` then the inferred key type of `element` is the inferred
+        type of `ek` in context `?` and the inferred value type of `element` is
+        the inferred type of `ev` in context `?`.
+
+    *   If `P` is `Map<Pk, Pv>` then the inferred key type of `element` is the
+        inferred type of `ek` in context `Pk` and the inferred value type of `element`
+        is the inferred type of `ev` in context `Pv`.
+
+*   If `element` is a `spreadElement` with expression `e1`:
+
+    *   If `P` is `?` then let `S` be the inferred type of `e1` in context `?`:
+
+        *   If `S` is a subtype of `Iterable<Object>` and not a subtype of
+            `Map<Object, Object>`, then the inferred set element type of
+            `element` is `T` where `T` is the type such that `Iterable<T>` is a
+            superinterface of `S` (the result of constraint matching for `X`
+            using the constraint `S <: Iterable<X>`).
+
+        *   If `S` is a subtype of `Map<Object, Object>` and not a subtype of
+            `Iterable<Object>`, then the inferred key type of `element` is `K`
+            and the inferred value type of `element` is `V`, where `K` and `V`
+            are the types such that `Map<K, V>` is a superinterface of `S` (the
+            result of constraint matching for `X` and `Y` using the constraint
+            `S <: Map<X, Y>`).
+
+        *   If `S` is `dynamic`, then the inferred set element type of `element`
+            is `dynamic`, the inferred key type of `element` is `dynamic`, and
+            the inferred value type of `element` is `dynamic`. *(We produce both
+            a set element type here, and a key/value pair here and rely on other
+            elements to disambiguate.)*
+
+        *   Otherwise it is an error (either because we cannot disambiguate,
+            such as with something that implements both Map and Iterable, or
+            because it is a spread of a non-spreadable type).
+
+    *   If `P` is `Set<Ps>` then let `S` be the inferred type of `e1` in context
+        `Iterable<Ps>`:
+
+        *   If `S` is a non-`Null` subtype of `Iterable<Object>`, then the
+            inferred set element type of `element` is `T` where `T` is the type
+            such that `Iterable<T>` is a superinterface of `S` (the result of
+            constraint matching for `X` using the constraint `Iterable<X> <:
+            S`).
+
+        *   If `S` is `dynamic`, then the inferred set element type of `element`
+            is `dynamic`.
+
+        *   Otherwise it is an error.
```

Comment:
It's a little trickier than that. We don't generally want to cast the *sequence* to an `Iterable<T>` because it may not be. It might be a super type. We allow that and then cast each element during iteration:

```dart
Iterable<num> nums = [1, 2, 3];
Set<int> ints = {...nums}; // Fine.
```

So in a case like this:

```dart
Object nums = <num>[1, 2, 3];
Set<int> ints = {...nums};
```

I'd rather not cast `nums` to `Iterable<int>`. It will fail, and not in a useful way, and not in a way that mirrors similar cases where the casts are only on the individual elements. But if we don't cast to the set's element type, what `Iterable<T>` would we cast it to? `Iterable<Object>`? That feels kind of arbitrary.

Since I don't think this is a common case anyway, I figured the simplest thing is to just make it a static error and let the user tell us what they want. The odds very good that a silent downcast is *not* what they want anyway. And this is consistent with the cases where we don't have a context type.

Reply:
Argument accepted :+1: